### PR TITLE
update Concept translator to map FHIR codes to concepts with any source mapping

### PR DESCRIPTION
**Pull request by mozzy11**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-501

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirConceptDaoImpl.java [hunk 1]
```diff
@@ -70,4 +70,30 @@ public Optional<Concept> getConceptWithSameAsMappingInSource(@Nonnull ConceptSou
...
+			criteria.add(eq("term.code", mappingCode).ignoreCase());
+		} else {
+			criteria.add(eq("term.code", mappingCode));
+		}
+		
+		criteria.add(eq("term.conceptSource", conceptSource));
+		criteria.addOrder(asc("concept.retired"));
+		criteria.setResultTransformer(DistinctRootEntityResultTransformer.INSTANCE);
+		
+		return Optional.ofNullable((Concept) criteria.uniqueResult());
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirConceptDaoImpl.java:97]: `uniqueResult()` throws an exception if you get more than one result which is, I think, not what we want here.

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ConceptTranslatorImpl.java [hunk 2]
```diff
@@ -103,6 +103,11 @@ public Concept toOpenmrsType(@Nonnull CodeableConcept concept) {
 							    coding.getCode());
 							if (c.isPresent()) {
 								return c.get();
+							} else {
+								c = conceptService.getConceptWithAnyMappingInSource(conceptSource.get(), coding.getCode());
+								if (c.isPresent()) {
+									return c.get();
+								}
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ConceptTranslatorImpl.java:110]: Do we need two database look-ups here? Maybe we can look into return a DTO that includes both the concept and the mapping type then change the logic around here: if we have a SAME-AS map use that, etc. as before?

**Review by ibacher — COMMENTED** [review]
Thanks @mozzy11! A couple of thoughts on what we're doing here.

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ConceptTranslatorImpl.java (hunk 2 above)
Comment (mozzy11) [reply api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ConceptTranslatorImpl.java]: Thanks @ibacher .
updated

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirConceptDaoImpl.java [hunk 3]
```diff
@@ -70,4 +72,30 @@ public Optional<Concept> getConceptWithSameAsMappingInSource(@Nonnull ConceptSou
 		
 		return Optional.ofNullable((Concept) criteria.uniqueResult());
 	}
+	
+	@Override
+	public List<Concept> getConceptsWithAnyMappingInSource(@Nonnull ConceptSource conceptSource,
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirConceptDaoImpl.java:77]: Can you add a test or two for this DAO method? Then I think this is good to merge.

Comment (mozzy11) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirConceptDaoImpl.java:77]: Thanks @ibacher , had forgoten to add tests for that one .
added

**Pull request merged**